curp/src/client_new/unary.rs

```diff
@@ -0,0 +1,698 @@
+        // thought it is the local error
+        let req =
+            FetchReadStateRequest::new(cmd, self.cluster_version().await).map_err(|ser_err| {
+                warn!("serializing error: {ser_err}");
+                CurpError::from(ser_err)
+            })?;
+        let timeout = self.config.wait_synced_timeout;
+        let state = self
+            .map_leader(|conn| async move { conn.fetch_read_state(req, timeout).await })
+            .await??
+            .into_inner()
+            .read_state
+            .unwrap_or_else(|| unreachable!("read_state must be set in fetch read state response"));
+        Ok(state)
+    }
+
+    /// Send fetch cluster requests to all servers
+    /// Note: The fetched cluster may still be outdated if `linearizable` is false
+    async fn fetch_cluster(&self, linearizable: bool) -> Result<FetchClusterResponse, CurpError> {
+        let timeout = self.config.wait_synced_timeout;
+        if !linearizable {
+            // firstly, try to fetch the local server
+            if let Some(connect) = <Unary<C> as ClientApi>::local_connect(self).await {
+                /// local timeout, in fact, local connect should only be bypassed, so the timeout maybe unused.
+                const FETCH_LOCAL_TIMEOUT: Duration = Duration::from_secs(1);
+
+                let resp = connect
+                    .fetch_cluster(FetchClusterRequest::default(), FETCH_LOCAL_TIMEOUT)
+                    .await
+                    .unwrap_or_else(|e| {
+                        unreachable!(
+                            "fetch cluster from local connect should never failed, err {e:?}"
+                        )
+                    })
+                    .into_inner();
+                return Ok(resp);
+            }
+        }
+        // then fetch the whole cluster
+        let (size, mut responses) = self
+            .for_each_server(|conn| async move {
+                (
+                    conn.id(),
+                    conn.fetch_cluster(FetchClusterRequest { linearizable }, timeout)
+                        .await
+                        .map(Response::into_inner),
+                )
+            })
+            .await;
+        let quorum = quorum(size);
+
+        let mut max_term = 0;
+        let mut res = None;
+        let mut ok_cnt = 0;
+        let mut err = None;
+
+        while let Some((id, resp)) = responses.next().await {
+            let inner = match resp {
+                Ok(r) => r,
+                Err(e) => {
+                    warn!("fetch cluster from {} failed, {:?}", id, e);
+                    if e.return_early() {
+                        return Err(e);
+                    }
+                    err = Some(e);
+                    continue;
+                }
+            };
+            // Ignore the response of a node that doesn't know who the leader is.
+            // some disconnected candidates may continue to increase term, disrupting the process below.
```

Comment:
Because of pre-vote, disconnected candidate will not increase term

Reply:
We could have a dual insurance mechanism.